mipengine/controller/algorithm_executor_nodes.py

```diff
@@ -70,13 +64,13 @@ def queue_run_udf(
         self,
         command_id: str,
         func_name: str,
-        positional_args: UDFPosArguments,
-        keyword_args: UDFKeyArguments,
+        positional_args: NodeUDFPosArguments,
+        keyword_args: NodeUDFKeyArguments,
     ) -> AsyncResult:
         pass
 
     @abstractmethod
-    def get_queued_udf_result(self, async_result: AsyncResult) -> List[TableName]:
+    def get_queued_udf_result(self, async_result: AsyncResult) -> List[NodeUDFDTO]:
```

Comment:
Ok you re right. I realized my confusion actually originated from the way the types are defined in 
```
class NodeUDFPosArguments(ImmutableBaseModel):
    args: List[Union[NodeLiteralDTO, NodeTableDTO, NodeSMPCDTO]]

class NodeUDFKeyArguments(ImmutableBaseModel):
    args: Dict[str, Union[NodeLiteralDTO, NodeTableDTO, NodeSMPCDTO]]

class NodeUDFResults(ImmutableBaseModel):
    results: List[Union[NodeLiteralDTO, NodeTableDTO, NodeSMPCDTO]]
```

Since you have defined a common parent class for `NodeLiteralDTO`, `NodeTableDTO` and `NodeSMPCDTO`, then it would be cleaner to define these as 
```
class NodeUDFPosArguments(ImmutableBaseModel):
    args: List[NodeUDFDTO]

class NodeUDFKeyArguments(ImmutableBaseModel):
    args: Dict[str, NodeUDFDTO]

class NodeUDFResults(ImmutableBaseModel):
    results: List[NodeUDFDTO]
```
declaring the type using a `Union` makes sense if there is a need to _unite_ classes that are unrelated, but `NodeLiteralDTO`, `NodeTableDTO` and `NodeSMPCDTO` are sibling classes ([ref](https://stackoverflow.com/questions/41790843/python-3-type-annotations-and-subclasses))

afterall, this is how this kind of object is already defined in other modules as well
```
algorithm_executor_nodes.py:73
def get_queued_udf_result(self, async_result: AsyncResult) -> List[NodeUDFDTO]:

algorithm_executor_nodes.py:190
self, command_id: str, func_name: str, positional_args: List[NodeUDFDTO]

algorithm_executor_nodes.py:249
def get_queued_udf_result(self, async_result: AsyncResult) -> List[NodeUDFDTO]:

algorithm_execution_tasks_handler.py:141
..... -> List[NodeUDFDTO]:
algorithm_execution_tasks_handler.py:460
.... -> List[NodeUDFDTO]:
```

Reply:
Ah okay, sure, will change the return type in NodeUDFPosArguments, NodeUDFKeyArguments and NodeUDFResults.